openlibrary/plugins/openlibrary/js/autocomplete.js

```diff
@@ -16,59 +16,92 @@ export default function($) {
             },
+            mustMatch: true,
+            formatMatch: function(item) { return item.name; }
         };
 
-        $(_this)
-            .autocomplete(ol_ac_opts.endpoint, $.extend(default_ac_opts, ac_opts))
-            .result(function(event, item) {
-                var $this;
 
-                $(`#${this.id}-key`).val(item.key);
-                $this = $(this);
+        /**
+         * Port of code in vendor/js/jquery-autocomplete removed in e91119b
+         * @param {string} value
+         * @param {string} term to highlight in value
+         * @return {string}
+         */
+        function highlight(value, term) {
+            return value.replace(
+                new RegExp(`(?![^&;]+;)(?!<[^<>]*)(${term.replace(/([\^$()[]\{\}\*\.\+\?\|\\])/gi, '$1')})(?![^<>]*>)(?![^&;]+;)`, 'gi'),
+                '<strong>$1</strong>'
+            );
+        }
+
+        $.widget('custom.autocompleteHTML', $.ui.autocomplete, {
+            _renderMenu($ul, items) {
+                $ul.addClass('ac_results');
+                items.forEach((item, i) => {
+                    $('<li>')
+                        .addClass(i % 2 ? 'ac-even' : 'ac-odd')
+                        .attr('data-value', item.value)
+                        .data('ui-autocomplete-item', item)
+                        .attr('aria-label', item.value)
+                        .html(item.label)
+                        .appendTo($ul);
+                });
+            }
+        });
+
+        const options = $.extend(default_ac_opts, ac_opts, {
+            source: function (q, response) {
+                const term = q.term;
+                return $.ajax({
+                    url: ol_ac_opts.endpoint,
+                    data: {
+                        q: term,
+                        limit: options.max,
+                        timestamp: new Date()
+                    }
+                }).then((results) => {
+                    response(
+                        results.map((r) => {
+                            return {
+                                label: highlight(options.formatItem(r), term),
+                                value: r.name
+                            };
+                        })
+                    );
 
-                //adding class directly is not working when tab is pressed. setTimeout seems to be working!
-                setTimeout(function() {
-                    $this.addClass('accept');
-                }, 0);
-            })
-            .nomatch(function(){
-                $(`#${this.id}-key`).val('');
-                $(this).addClass('reject');
-            })
-            .keypress(function() {
+                    // When no results if callback is defined, append a create new entry
+                    if (!results.length &&
+                        (
+                            ol_ac_opts.addnew ||
+                            (ol_ac_opts.addnew && ol_ac_opts.addnew(term))
```

Comment:
I don't quite understand what's happening with this expression.  If `addnew` is defined, won't this always return true regardless of what `ol_ac_opts.addnew(term)` returns?

Reply:
I don't understand either :) Generally when I port code, I'm pretty conservative to retain existing logic where ever possible. This is the code that was originally on L41 just in a different place:

https://github.com/internetarchive/openlibrary/blob/a61250b4ade345430e406d99eb04ae102713899a/openlibrary/plugins/openlibrary/js/autocomplete.js#L41

The `parse` option is no longer supported for the jquery ui autocomplete, so this instead needs to run as part of the `source` option.

Once we've got onto jquery UI that means we can push ahead with upgrading the jquery version, which will mean writing lots of unit tests for things like this.

That make sense?

Reply:
And now as I write this I see I missed the boolean check when porting which does make this make more sense :)

Reply:
Now:
```
ol_ac_opts.addnew === true ||
                            (ol_ac_opts.addnew && ol_ac_opts.addnew(term))
                            ```